lib/solarwinds_apm/support/resource_detector.rb

```diff
@@ -0,0 +1,180 @@
+require 'securerandom'
+
+module SolarWindsAPM
+  # ResourceDetector
+  # Usage:
+  #   require 'opentelemetry/sdk'
+  #   require 'opentelemetry/resource/detector'
+  #   OpenTelemetry::SDK.configure do |c|
+  #     c.resource = SolarWindsAPM::ResourceDetector.detect
+  #   end
+  module ResourceDetector
+    K8S_PODNAME_PATH = '/etc/hostname'
+    K8S_NAMESPACE_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+    K8S_NAMESPACE_PATH_WIN = 'C:\\var\\run\\secrets\\kubernetes.io\\serviceaccount\\namespace'
+    K8S_MOUNTINFO_FILE = '/proc/self/mountinfo'
+    UID_REGEX = /[0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12}/i
+
+    SW_K8S_NAMESPACE_ENV = 'SW_K8S_POD_NAMESPACE'
+    SW_K8S_UID_ENV = 'SW_K8S_POD_UID'
+    SW_K8S_NAME_ENV = 'SW_K8S_POD_NAME'
+
+    UAMS_CLIENT_PATH = '/opt/solarwinds/uamsclient/var/uamsclientid'
+    UAMS_CLIENT_PATH_WIN = 'C:\\ProgramData\\SolarWinds\\UAMSClient\\uamsclientid'
+    UAMS_CLIENT_URL = 'http://127.0.0.1:2113/info/uamsclient'
+    UAMS_CLIENT_ID_FIELD = 'uamsclient_id'
+
+    def self.detect
+      attributes = ::OpenTelemetry::SDK::Resources::Resource.create({})
+      attributes = attributes.merge(detect_uams_client_id)
+      attributes = attributes.merge(detect_k8s_attributes)
+      attributes.merge(from_upstream_detector)
+    end
+
+    def self.detect_uams_client_id
+      umas_client_final_path = windows? ? UAMS_CLIENT_PATH_WIN : UAMS_CLIENT_PATH
+      uams_client_id = nil
+      begin
+        uams_client_id = File.read(umas_client_final_path).strip
+      rescue StandardError => e
+        SolarWindsAPM.logger.debug "#{self.class}/#{__method__}] umas file retrieve error #{e.message}."
+      end
+
+      if uams_client_id.nil?
+        begin
+          url = URI(UAMS_CLIENT_URL)
+
+          response = nil
+          ::OpenTelemetry::Common::Utilities.untraced do
+            http = Net::HTTP.new(url.host, url.port)
+            request = Net::HTTP::Get.new(url)
+            response = http.request(request)
+          end
+
+          raise 'Response returned non-200 status code' unless response&.code.to_i == 200
+
+          uams_metadata = JSON.parse(response.body)
+          uams_client_id = uams_metadata&.fetch(UAMS_CLIENT_ID_FIELD)
+        rescue StandardError => e
+          SolarWindsAPM.logger.debug "#{self.class}/#{__method__}] umas api retrieve error #{e.message}."
+        end
+      end
+
+      resource_attributes = {
+        'sw.uams.client.id' => uams_client_id,
+        'host.id' => uams_client_id
+      }
+
+      SolarWindsAPM.logger.debug "#{self.class}/#{__method__}] retrieved resource_attributes: #{resource_attributes.inspect}."
+      ::OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+    rescue StandardError => e
+      SolarWindsAPM.logger.debug "#{self.class}/#{__method__}] detect_uams_client_id failed. Error: #{e.message}."
+      ::OpenTelemetry::SDK::Resources::Resource.create({})
+    end
+
+    def self.detect_k8s_attributes
```

Comment:
How about a debug log line like some of the other detectors to indicate skipping because not running in k8s?

Reply:
Updated.